src/ast/ast.c

```diff
@@ -72,6 +73,24 @@ static void _AST_LimitResults(AST *ast, const cypher_astnode_t *root_clause,
 	}
 }
 
+static void _extract_params(const cypher_astnode_t *statement) {
+	rax *params = NULL;
+	uint noptions =  cypher_ast_statement_noptions(statement);
+	for(uint i = 0; i < noptions; i++) {
+		const cypher_astnode_t *option = cypher_ast_statement_get_option(statement, i);
+		uint nparams = cypher_ast_cypher_option_nparams(option);
+		for(uint j = 0; j < nparams; j++) {
+			const cypher_astnode_t *param = cypher_ast_cypher_option_get_param(option, j);
+			const char *paramName = cypher_ast_string_get_value(cypher_ast_cypher_option_param_get_name(param));
+			const cypher_astnode_t *paramValue = cypher_ast_cypher_option_param_get_value(param);
+			if(!params) params = raxNew();
+			AR_ExpNode *exp = AR_EXP_FromExpression(paramValue);
```

Comment:
Double free on:
`CYPHER param=[0, 1, 2] RETURN $param`
```
==5553== Invalid read of size 4
==5553==    at 0x84BDDE6: array_len (arr.h:206)
==5553==    by 0x84BDF70: SIArray_Length (array.c:29)
==5553==    by 0x84BE353: SIArray_Free (array.c:77)
==5553==    by 0x84886C8: SIValue_Free (value.c:595)
==5553==    by 0x8498FCF: AR_EXP_Free (arithmetic_expression.c:594)
==5553==    by 0x854FA2C: _FreeParamAnnotationCallback (annotate_params.c:19)
==5553==    by 0x8B427D1: cp_release_annotation (in /home/tower-linux/dev/redis/graph/src/redisgraph.so)
==5553==    by 0x8B4223F: cypher_ast_annotation_context_free (in /home/tower-linux/dev/redis/graph/src/redisgraph.so)
==5553==    by 0x854D03F: AST_Free (ast.c:372)
==5553==    by 0x84B8993: Graph_Query (cmd_query.c:133)
==5553==    by 0x856D96A: thread_do (thpool.c:337)
==5553==    by 0x55ED6DA: start_thread (pthread_create.c:463)
==5553==  Address 0x117fdc50 is 0 bytes inside a block of size 60 free'd
==5553==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5553==    by 0x84BDE13: array_free (arr.h:230)
==5553==    by 0x84BE3A0: SIArray_Free (array.c:82)
==5553==    by 0x84886C8: SIValue_Free (value.c:595)
==5553==    by 0x8498FCF: AR_EXP_Free (arithmetic_expression.c:594)
==5553==    by 0x84ED0D9: ProjectFree (op_project.c:81)
==5553==    by 0x84F3477: OpBase_Free (op.c:129)
==5553==    by 0x84CC41A: _ExecutionPlan_FreeOperations (execution_plan.c:937)
==5553==    by 0x84CC3FE: _ExecutionPlan_FreeOperations (execution_plan.c:935)
==5553==    by 0x84CC55B: ExecutionPlan_Free (execution_plan.c:961)
==5553==    by 0x84B88DE: Graph_Query (cmd_query.c:113)
==5553==    by 0x856D96A: thread_do (thpool.c:337)
==5553==  Block was alloc'd at
==5553==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==5553==    by 0x84BDCE9: array_new_sz (arr.h:68)
==5553==    by 0x84BDE35: SIArray_New (array.c:8)
==5553==    by 0x8486E49: SI_Array (value.c:79)
==5553==    by 0x84A38A0: AR_TOLIST (list_funcs.c:15)
==5553==    by 0x849819A: _AR_EXP_EvaluateFunctionCall (arithmetic_expression.c:300)
==5553==    by 0x8498474: _AR_EXP_Evaluate (arithmetic_expression.c:371)
==5553==    by 0x8498541: AR_EXP_Evaluate (arithmetic_expression.c:392)
==5553==    by 0x8497D2A: AR_EXP_ReduceToScalar (arithmetic_expression.c:195)
==5553==    by 0x854AB29: AR_EXP_FromExpression (ast_build_ar_exp.c:414)
==5553==    by 0x854C274: _extract_params (ast.c:87)
==5553==    by 0x854C880: AST_Build (ast.c:220)
```
I do favor this approach, but I'm not sure if this is a simple issue to solve - thoughts?

Reply:
Fixed by https://github.com/RedisGraph/RedisGraph/pull/731/commits/012943676883fd1ef0a68cd741260299bdb47518#diff-d2446fec994a31524307c80826c0ec02R46

I see no new leaks; we're good!